src/eos-update-server.c

```diff
@@ -263,6 +264,58 @@ data_generate_config (Data *data,
   return TRUE;
 }
 
+static gchar *
+quit_file_name (void)
+{
+  return eos_updater_dup_envvar_or ("EOS_UPDATER_TEST_UPDATE_SERVER_QUIT_FILE",
+                                     NULL);
+}
+
+static EosQuitFileCheckResult
+check_and_quit (gpointer data_ptr)
+{
+  Data *data = data_ptr;
+  FILE *log;
+  g_autofree gchar *str = NULL;
+
+  log = fopen ("eos-update-server.log", "a");
+  str = g_strdup_printf ("requests pending: %u\n",
+                         data->async_requests_pending);
+  fwrite (str, 1, strlen (str), log);
+  fclose (log);
+  if (data->async_requests_pending > 0)
+    return EOS_QUIT_FILE_KEEP_CHECKING;
+
+  g_main_loop_quit (data->loop);
+  return EOS_QUIT_FILE_QUIT;
+}
+
+static gboolean
+maybe_setup_quit_file (Data *data,
+                       GError **error)
+{
+  g_autofree gchar *filename = quit_file_name ();
+  g_autoptr(EosQuitFile) quit_file = NULL;
+  FILE *log;
+
+  log = fopen ("eos-update-server.log", "w");
+  fclose (log);
```

Comment:
You can probably call `fclose` on the result of `fopen` if there's no need to keep the variable around.

(Though I wonder if there's a better function to create a file if it doesn't exist?)